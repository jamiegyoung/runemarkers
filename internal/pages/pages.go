package pages

import (
	"fmt"
	"github.com/jamiegyoung/runemarkers-go/internal/entities"
	"github.com/jamiegyoung/runemarkers-go/internal/logger"
	"github.com/jamiegyoung/runemarkers-go/internal/pageio"
	"path/filepath"
	"sync"
)

var log = logger.New("pages")

type Page struct {
	data map[string]interface{}
}

func (p Page) Data() map[string]interface{} {
	return p.data
}

func NewPage(data map[string]interface{}) Page {
	return Page{
		data: data,
	}
}

func GeneratePages(destination string, entities []*entities.Entity) error {
	paths, err := filepath.Glob("templates/pages/*.tmpl")
	if err != nil {
		return err
	}

	log("Found " + fmt.Sprint(len(paths)) + " page(s)")

	data := NewPage(map[string]interface{}{"Entities": entities})

	var wg sync.WaitGroup
	errc := make(chan error, len(paths))

	for _, path := range paths {
		wg.Add(1)

		go func(path string) {
			defer wg.Done()

			page, err := pageio.ReadPageString(path)
			if err != nil {
				errc <- err
				return
			}

			// create the output directory if it doesn't exist
			output, err := pageio.CreateOutFile(destination, path)
			defer output.Close()
			if err != nil {
				errc <- err
				return
			}

			name := filepath.Base(path)

			err = pageio.RenderHtml(name, page, output, &data)
			errc <- err
		}(path)
	}

	wg.Wait()
	close(errc)

	for err := range errc {
		if err != nil {
			return err
		}
	}

	return nil
}
