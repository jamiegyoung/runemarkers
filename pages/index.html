<!DOCTYPE html>
<html lang="en">
  <head>
    <title>{{.Title}}</title>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link href="css/style.css" rel="stylesheet">
  </head>
  <body>
    <h1>This is a test thing</h1>
    {{ range .Entities }}
      {{ template "entity" . }}
    {{ end }}
    {{.Body}}
  </body>
</html>
