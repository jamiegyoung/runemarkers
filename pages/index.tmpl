<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link href="css/style.css" rel="stylesheet">
  </head>
  <body>
    {{ template "navbar" }}
    <div class="main-container">
      <p class="help-text">Make sure the "Ground Markers" plugin is enabled on RuneLite with "Remember color per tile" enabled. Copy the tile markers to your clipboard, right-click the world map button in-game and select "import".</p>
      <div class="inner-list-container">
        {{ range .Entities }}
          {{- template "entity_card" . -}}
        {{ end }}
      </div>
    </div>
  </body>
</html>

<style>
html,
body {
  padding: 0;
  margin: 0;
  font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Oxygen,
    Ubuntu, Cantarell, Fira Sans, Droid Sans, Helvetica Neue, sans-serif;
  --color-primary: #e5e53a;
  --color-primary-light: #f1f196;
  --color-background: #20201f;
  --color-foreground: #292927;
  --color-text: #f5f3f5;
  --color-subtle-text: #c9c9c9;
  --color-card: #3b3b37;
  --color-card-hover: #42423e;
  --color-card-hover-light: #55554a;
  background-color: var(--color-background);
  color: var(--color-text);
}

.main-container {
  margin-top: 4rem;
  display: flex;
  flex-direction: column;
  align-items: center;
}

.help-text {
  padding: 0;
  color: rgb(199, 199, 199);
  text-align: center;
  min-width: 17rem;
  margin-bottom: 1rem;
  max-width: 40rem;
  width: 80%;
}

.inner-list-container {
  min-width: 17rem;
  margin-bottom: 0.75rem;
  max-width: 37rem;
  width: 80%;
  border-radius: 0.75rem;
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 1.25rem 0;
  background-color: var(--color-foreground);
  box-shadow: 0 .25rem .75rem rgba(0, 0, 0, 0.25), 0 .25rem .75rem rgba(0, 0, 0, 0.3);
  overflow-y: auto;
  overflow-x: hidden;
}

@media (max-width: 900px) {
  .inner-list-container {
    width: 95%;
    height: auto;
  }
}
</style>
