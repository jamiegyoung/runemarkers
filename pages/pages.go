package pages

import (
	"fmt"
	"github.com/jamiegyoung/runemarkers-go/entities"
	"github.com/jamiegyoung/runemarkers-go/libs"
	"github.com/jamiegyoung/runemarkers-go/logger"
	"github.com/jamiegyoung/runemarkers-go/pageio"
	"path/filepath"
	"sync"
)

var log = logger.New("pages")

const pages_glob = "pages/*.tmpl"

type IndexPage struct {
	pageio.Page
	Entities []*entities.Entity
}

func (p *IndexPage) Data() map[string]interface{} {
	return map[string]interface{}{
		"Entities": p.Entities,
	}
}

func GeneratePages(output_path string, found_entities []*entities.Entity) {
	pages_paths, err := filepath.Glob(pages_glob)
	if err != nil {
		panic(err)
	}

	log("Found " + fmt.Sprint(len(pages_paths)) + " page(s)")

	page_data := IndexPage{
		Entities: found_entities,
	}

	var wg sync.WaitGroup

	for _, page_path := range pages_paths {
		wg.Add(1)

		go func(page_path string) {
			defer wg.Done()

			page_string, err := pageio.ReadPageString(page_path)
			if err != nil {
				panic(err)
			}

			// create the output directory if it doesn't exist
			out_file := pageio.CreateOutFile(output_path, page_path)
			defer out_file.Close()

			page_name := filepath.Base(page_path)

			pageio.RenderPage(page_name, page_string, out_file, &page_data)
		}(page_path)
	}

	wg.Wait()

	err = libs.CopyLibs(output_path + "/js")
	if err != nil {
		panic(err)
	}

	log("Done!")
}
