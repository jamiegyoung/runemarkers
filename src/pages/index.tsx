import Fuse from 'fuse.js';
import { TileEntity } from '@/types';
import {
  Dispatch,
  SetStateAction,
  createContext,
  useEffect,
  useState,
} from 'react';
import TileEntities from '@/components/organisms/TileEntities';
import NavBar from '@/components/molecules/NavBar';
import styles from '@/styles/Home.module.css';
import { getTileData } from '@/api/tiles';
import ContributionFooter from '@/components/atoms/ContributionFooter';
import { NextSeo } from 'next-seo';
import { defaultImages } from '@/api/seoOptions';

const fuseOptions = {
  // the keys of the objects to search
  keys: [
    { name: `fullName`, weight: 0.7 },
    { name: `fullAltName`, weight: 0.5 },
    { name: `tags`, weight: 0.3 },
  ],
  threshold: 0.4,
};

export function getStaticProps() {
  return {
    props: { tileData: getTileData() },
  };
}

export const SearchContext = createContext<
  [string | undefined, Dispatch<SetStateAction<string>> | undefined]
>([undefined, undefined]);

export default function Home({ tileData }: { tileData: TileEntity[] }) {
  const [searchVal, setSearchVal] = useState(``);
  const [searchRes, setSearchRes] = useState<TileEntity[]>(tileData);

  useEffect(() => {
    const fuse = new Fuse(tileData, fuseOptions);
    if (searchVal.length > 0) {
      setSearchRes(fuse.search(searchVal).map((res) => res.item));
      return;
    }
    setSearchRes(tileData);
  }, [searchVal, tileData]);
  console.log(
    `
P########B5?:      ?BBB5              .PBBB~
&@@@#BBB#&@@@&!    5&&&&B.           :&&&&&7
&@@@:     ^&@@@^   5&&&&&#.         ^&&&&&&7
&@@@:      5@@@7   5&&#5&&&^       !&&&Y&&&7
&@@@:     !@@@#    5&&& 7&&&~     7&&&^.&&&7
&@@@#BBB#&@@#?     5&&&  !&&&7   J&&&^ .&&&7
&@@@#GB@@@@J       5&&&   ^&&&J 5&&#.  .&&&7
&@@@:  .G@@@P      5&&&    :&&&B&&#.   .&&&7
&@@@:    ?@@@&^    5&&&     .#&&&B     .&&&7
&@@@^     :&@@@Y   5&&&       7J7      .&&&7
G&#&:       5##&5  ?##G                .BB#~

Made with \u2764 by jamgyo`,
  );
  return (
    <>
      <NextSeo
        openGraph={{
          images: defaultImages,
        }}
      />
      <div>
        <main className={styles.main}>
          <SearchContext.Provider value={[searchVal, setSearchVal]}>
            <NavBar />
            <TileEntities list={searchRes} />
            <ContributionFooter />
          </SearchContext.Provider>
        </main>
      </div>
    </>
  );
}
