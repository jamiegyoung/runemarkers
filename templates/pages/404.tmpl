<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <script src="js/fuse.min.js"></script>
  <script src="js/htmx.min.js"></script>
  <link rel="icon" href="assets/favicon.ico" type="image/x-icon">
  <title>RuneMarkers - A Collection of RuneLite Tile Markers</title>
  <meta name="description" content="A collection of tile markers for RuneLite. Find and import tile markers for different Oldschool RuneScape activities.">
  <meta name="twitter:card" content="summary_large_image">
  <meta name="twitter:site" content="@runemarkers">
  <meta name="twitter:creator" content="@jamgyo">
  <meta property="og:title" content="RuneMarkers - A Collection of RuneLite Tile Markers">
  <meta property="og:description" content="A collection of tile markers for RuneLite. Find and import tile markers for different Oldschool RuneScape activities.">
  <meta property="og:url" content="https://runemarkers.net">
  <meta property="og:type" content="website">
  <meta property="og:locale" content="en_US">
  <meta property="og:site_name" content="RuneMarkers">
  <meta name="author" content="RuneMarkers">
  <meta charset="utf-8">
  <meta name="viewport" content="initial-scale=1.0, width=device-width">
  <meta name="theme-color" content="#262624">
  <meta name="robots" content="index,follow">
  <meta property="og:image" content="/assets/logo-512-background.png">
  <meta property="og:image:alt" content="RuneMarkers - A Collection of RuneLite Tile Markers">
  <meta property="og:image:width" content="512">
  <meta property="og:image:height" content="512">
  <meta name="next-head-count" content="40">
  <link rel="apple-touch-icon" sizes="180x180" href="/assets/apple-touch-icon.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/assets/favicon-32x32.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/assets/favicon-16x16.png">
  {{ template "styles" .}}

  <style>
    html,
    body {
      box-sizing: border-box;
      padding: 0;
      margin: 0;
      font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Oxygen,
        Ubuntu, Cantarell, Fira Sans, Droid Sans, Helvetica Neue, sans-serif;
      --color-primary: #e5e53a;
      --color-primary-light: #f1f196;
      --color-background: #20201f;
      --color-foreground: #292927;
      --color-text: #f5f3f5;
      --color-subtle-text: #cccccc;
      --color-card: #3b3b37;
      --color-card-hover: #42423e;
      --color-card-hover-light: #55554a;
      background-color: var(--color-background);
      color: var(--color-text);
      height: calc(100% - 4rem);
    }

    *, *::before, *::after {
      box-sizing: inherit;
    }

    a {
      color: var(--color-primary);
      transition: color .2s ease-out;
    }

    a:hover {
      color: var(--color-primary-light);
    }

    .main-container {
      margin-top: 4rem;
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: center;
      height: 100%;
    }
  </style>
</head>

<body>
  {{ template "navbar" }}
  <main class="main-container">
    <h1><span style="color: var(--color-primary);">404</span> - Page Not Found</h1>
    <a href="/">Go back to the homepage</a>
  </main>
</body>

</html>

