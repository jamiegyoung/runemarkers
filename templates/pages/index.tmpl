<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <script src="js/fuse.js"></script>
  <script src="js/htmx.min.js"></script>
</head>

<body>
  {{ template "navbar" }}
  <main class="main-container">
    <p class="help-text">Make sure the "Ground Markers" plugin is enabled on RuneLite with "Remember color per tile"
      enabled. Copy the tile markers to your clipboard, right-click the world map button in-game and select "import".
    </p>
    {{ template "search" . }}
    {{ template "entity_list" . }}
  </main>
</body>

</html>

<script>
  htmx.config.selfRequestsOnly = true;
</script>

{{ template "styles" .}}

<style>
  html,
  body {
    padding: 0;
    margin: 0;
    font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Oxygen,
      Ubuntu, Cantarell, Fira Sans, Droid Sans, Helvetica Neue, sans-serif;
    --color-primary: #e5e53a;
    --color-primary-light: #f1f196;
    --color-background: #20201f;
    --color-foreground: #292927;
    --color-text: #f5f3f5;
    --color-subtle-text: #cccccc;
    --color-card: #3b3b37;
    --color-card-hover: #42423e;
    --color-card-hover-light: #55554a;
    background-color: var(--color-background);
    color: var(--color-text);
  }

  .main-container {
    margin-top: 4rem;
    display: flex;
    flex-direction: column;
    align-items: center;
  }

  .help-text {
    padding: 0;
    color: var(--color-subtle-text);
    text-align: center;
    min-width: 17rem;
    margin-bottom: 1rem;
    max-width: 40rem;
    width: 80%;
  }
</style>

