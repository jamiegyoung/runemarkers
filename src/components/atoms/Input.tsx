import { SearchContext } from '@/pages';
import styles from '@/components/atoms/Input.module.css';
import { useContext, useEffect, useRef, useState } from 'react';
import useDebounce from '@/hooks/useDebounce';

export default function Input() {
  const [searching, setSearching] = useState(false);
  const [searchVal, setSearchVal] = useContext(SearchContext);
  const inputRef = useRef<HTMLInputElement>(null);
  const [inputWidth, setInputWidth] = useState(0);

  const updateWidth = () => {
    if (!inputRef.current) return;
    setInputWidth(inputRef.current.offsetWidth);
  };

  const updateWidthDebounced = useDebounce(updateWidth, 100);

  useEffect(() => {
    updateWidth();
    window.addEventListener(`resize`, updateWidthDebounced);
  });

  return (
    <>
      <input
        ref={inputRef}
        type="text"
        placeholder={
          searching
            ? ``
            : inputWidth > 230
              ? `type here to search`
              : `search tiles`
        }
        className={[styles.input, searching ? styles.active : null].join(` `)}
        value={searchVal}
        onFocus={() => setSearching(true)}
        onBlur={() => {
          if (searchVal === ``) {
            setSearching(false);
          }
        }}
        onChange={(e) => {
          if (setSearchVal) {
            setSearchVal(e.target.value);
          }
        }}
      />
      <button
        className={[styles.clearButton, searching ? styles.active : null].join(
          ` `,
        )}
        onClick={() => {
          if (setSearchVal) {
            setSearching(false);
            setSearchVal(``);
          }
        }}
      >
        <svg xmlns="http://www.w3.org/2000/svg" height="24" width="24">
          <path d="M6.4 19.45 4.55 17.6l5.6-5.6-5.6-5.6L6.4 4.55l5.6 5.6 5.6-5.6 1.85 1.85-5.6 5.6 5.6 5.6-1.85 1.85-5.6-5.6Z" />
        </svg>
      </button>
    </>
  );
}
