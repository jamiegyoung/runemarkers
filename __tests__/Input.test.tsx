import { SearchContext } from '@/pages';
import '@testing-library/jest-dom';
import { act, fireEvent, render } from '@testing-library/react';
import { Dispatch, SetStateAction } from 'react';
import Input from '@/components/atoms/Input';

const testInputWithContext = (
  searchVal: string | undefined,
  setSearchVal: Dispatch<SetStateAction<string>>,
) => (
  <SearchContext.Provider value={[searchVal, setSearchVal]}>
    <div>
      <Input />
      <button></button>
    </div>
  </SearchContext.Provider>
);

describe(`Input`, () => {
  it(`should render correctly`, () => {
    const { container } = render(<Input />);
    expect(container).toMatchSnapshot();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  const mockSetSearchVal = jest.fn();

  jest.mock(`react`, () => ({
    useState: (initial: string) => [initial, mockSetSearchVal],
  }));

  it(`should expand on focus`, () => {
    const { container } = render(testInputWithContext(``, mockSetSearchVal));

    const input = container.querySelector(`input`);
    if (!input) {
      throw new Error(`No input found`);
    }
    act(() => {
      fireEvent.focus(input);
    });
    expect(input.placeholder).toBe(``);
  });

  it('should show the full placeholder when it is a sufficient width', async () => {
    Object.defineProperty(HTMLElement.prototype, 'offsetWidth', {
      configurable: true,
      value: 250, // Set this to a value greater than 230
    });

    const { container } = render(testInputWithContext(``, mockSetSearchVal));

    const input = container.querySelector(`input`);
    if (!input) {
      throw new Error(`No input found`);
    }

    await act(async () => {
      // Trigger the event
      window.dispatchEvent(new Event(`resize`));
    })

    expect(input.placeholder).toBe(`type here to search`);
  });

  it('should show the partial placeholder when it is a insufficient width', async () => {
    Object.defineProperty(HTMLElement.prototype, 'offsetWidth', {
      configurable: true,
      value: 200, // Set this to a value greater than 230
    });

    const { container } = render(testInputWithContext(``, mockSetSearchVal));

    const input = container.querySelector(`input`);
    if (!input) {
      throw new Error(`No input found`);
    }

    await act(async () => {
      // Trigger the event
      window.dispatchEvent(new Event(`resize`));
    })

    expect(input.placeholder).toBe(`search tiles`);
  });

  it(`should shrink on blur`, () => {
    const { container } = render(testInputWithContext(``, mockSetSearchVal));

    const input = container.querySelector(`input`);
    if (!input) {
      throw new Error(`No input found`);
    }

    act(() => {
      fireEvent.focus(input);
    });
    expect(input.placeholder).toBe(``);
    act(() => {
      input.value = ``;
      fireEvent.blur(input);
    });
    expect(input.placeholder).toBe(`search tiles`);
  });

  it(`should update the search value on change`, () => {
    const { container } = render(testInputWithContext(``, mockSetSearchVal));

    const input = container.querySelector(`input`);
    if (!input) {
      throw new Error(`No input found`);
    }
    act(() => {
      fireEvent.focus(input);
    });
    expect(input.placeholder).toBe(``);
    act(() => {
      fireEvent.change(input, { target: { value: `test` } });
    });
    expect(mockSetSearchVal).toHaveBeenCalledTimes(1);
    expect(mockSetSearchVal).toHaveBeenCalledWith(`test`);
  });

  it(`should clear the search value when the clear button is clicked`, () => {
    const { container } = render(testInputWithContext(``, mockSetSearchVal));

    const input = container.querySelector(`input`);
    if (!input) {
      throw new Error(`No input found`);
    }

    act(() => {
      fireEvent.change(input, { target: { value: `test` } });
    });

    expect(mockSetSearchVal).toHaveBeenCalledTimes(1);
    expect(mockSetSearchVal).toHaveBeenCalledWith(`test`);
    const clearButton = container.querySelector(`.clearButton`);
    if (!clearButton) {
      throw new Error(`No clear button found`);
    }
    act(() => {
      fireEvent.click(clearButton);
    });
    expect(mockSetSearchVal).toHaveBeenCalledTimes(2);
    expect(mockSetSearchVal).toHaveBeenCalledWith(``);
  });
});
