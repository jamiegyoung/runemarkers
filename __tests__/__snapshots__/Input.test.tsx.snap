// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render correctly 1`] = `
<div>
  <input
    class="input "
    placeholder="search tiles"
    type="text"
    value=""
  />
  <button
    class="clearButton "
  >
    <svg
      height="24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M6.4 19.45 4.55 17.6l5.6-5.6-5.6-5.6L6.4 4.55l5.6 5.6 5.6-5.6 1.85 1.85-5.6 5.6 5.6 5.6-1.85 1.85-5.6-5.6Z"
      />
    </svg>
  </button>
</div>
`;
